### PR TITLE
Add guide to writing plugins

### LINKED ISSUE #106: feat: Add article about how to write a plugin
This should link to a GitHub template repo that contains a basic setup.

### PR DESCRIPTION
**Pull request by KnorpelSenf**
Based on https://github.com/Amir-Zouerami/plugin-template

Closes #106

### CODE REVIEW

**Comment by KnorpelSenf** [comment]
I changed most lines, but almost none of the content. The only real difference is that I stripped down the middleware example a bit more.

There's currently an overlap between what we say in the “About Plugins” section and what we say in the early parts of this new guide. We need to figure out how those two documents are supporting each in other, instead of repeating each other.

Moreover, we currently have all the content of the guide in the README of the repo, too. This is not optimal, because while seemingly identical, the documents are going to diverge over time. We should figure out what to do with the README. I suggest to strip it down by a lot, and only reference this guide from now on. That's what is done with most other plugins etc.

**Comment by Amir-Zouerami** [comment]
I suggest mixing the two pages together to get everything in one place because the "about plugins" page already talks about how to submit a plugin. I can cut down the README but what do you suggest to keep in? I would think that we keep "rules of contribution" and then say "go to this page for instructions on how to use this template repository". Just a suggestion.

**Comment by KnorpelSenf** [comment]
> I suggest mixing the two pages together to get everything in one place

I disagree. There are two kinds of readers, plugin users (includes newcomers who don't have a clue) and people who know the framework to some extent and want to contribute a plugin. It's much easier to target those two groups individually.

It would be easier to take out the first section of the new guide and rather link to the About Plugins section, saying it is a prerequisite to writing plugins. (Maybe we should actually also link to middleware and transformers in the respective sections, just to make sure people know what they need to know before writing a plugin.)

> I would think that we keep "rules of contribution" and then say "go to this page for instructions on how to use this template repository".

That sounds good! Let's do it that way.

**Comment by Amir-Zouerami** [comment]
The README is trimmed down.

> I disagree. There are two kinds of readers, plugin users (includes newcomers who don't have a clue) and people who know the framework to some extent and want to contribute a plugin. It's much easier to target those two groups individually.

The developing part of plugins is kind of tied to understanding them. How about we transfer the following sections to the "about the plugins":
- What Is a Plugin?
- Types of Plugins in grammY (minus the stuff about how "your plugin should return this and that").

Since we link the words "middleware" and "transformer", readers can click on and check what they are if they're unfamiliar with them. Then in the hitchhiker's section we can add a summary of two types of plugins (link them to "about plugins") and carry on with "Rules of Contribution".

**Comment by KnorpelSenf** [comment]
I like that, that sounds good. That way, we can have everyone read the about section, and those who are interested in writing a plugin can read on.

Do you want to do the restructuring?

**Comment by Amir-Zouerami** [comment]
> Do you want to do the restructuring?

I will. I'll let you know when it's finished.

**Comment by Amir-Zouerami** [comment]
> Nice one, did not think about this

 Some guy name Dunkan in the Telegram group gave me the idea. All credits to him.

**Comment by KnorpelSenf** [comment]
Thanks @dcdunkan :+1:

**Comment by Amir-Zouerami** [comment]
I restructured [README ](https://github.com/grammyjs/website/commit/fe98a7d5bbb3736e2968c1e5dee34433749bfa72) and the [guide](https://github.com/grammyjs/website/commit/9a8b1877aba188b2a1a7a2d9429616a53033c847). Let me know what you think.

**Comment by KnorpelSenf** [comment]
Thanks a bunch! I generally agree with this structure. There are a few minor refinements I'd like to have here and there, but I'll just rephrase those parts myself (I think that would take as much time as writing down here what I have in mind).

**Comment by Amir-Zouerami** [comment]
great 👍

**Review by Amir-Zouerami — APPROVED** [review]
Well done on brushing the details especially making deno mandatory for official plugins. I wanted to do that but thought that you might not like it. Go ahead and merge whenever you're ready.

Nice job everyone who helped👍🏻

**Comment by KnorpelSenf** [comment]
> making deno mandatory for official plugins

This has already been required by https://github.com/grammyjs/grammY/blob/main/CONTRIBUTING.md#how-to-contribute for a long time, it is just a lot more visible now :)

We need to wait for translations.

**Review by roj1512 — APPROVED** [review]

File: site/docs/plugins/guide.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+      await ctx.reply(`I'm not talking to you! You don't care about ${str}!`);
+    }
+  };
+}
+```
+
+Now, it can be used in a real bot:
+
+```ts
+import { onlyAccept } from "./testing.ts";
```
Comment (WingLim) [inline site/docs/plugins/guide.md:82]: Here should be `plugin.ts`.
```suggestion
import { onlyAccept } from "./plugin.ts";
```

File: site/docs/plugins/guide.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+// Importing the types from grammY (we re-export them in `deps.deno.ts`).
+import type { Context, Middleware, NextFunction } from "./deps.deno.ts";
+
+// Your plugin can have one main function that creates middleware
+export function onlyAccept<C extends Context>(str: string): Middleware<C> {
+  // Create and return middleware.
+  return async (ctx, next) => {
+    // Get first name of user.
+    const name = ctx.from?.first_name;
+    // Let through all matching updates
```
Comment (WingLim) [inline site/docs/plugins/guide.md:67]: Add period.
```suggestion
    // Let through all matching updates.
```

File: site/docs/plugins/guide.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+This means that while your bot is sending a file, users will automatically see “_sending file…_” as status.
+Pretty cool, right?
+
+```ts
+// plugin.ts
+import type { Transformer } from "./deps.deno.ts";
+
+// Main plugin function
+export function autoChatAction(): Transformer {
+  // Create and return a transformer function
```
Comment (WingLim) [inline site/docs/plugins/guide.md:111]: Add period.
```suggestion
  // Create and return a transformer function.
```

File: site/docs/plugins/guide.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  return async (prev, method, payload, signal) => {
+    // Save the handle of the set interval so we can clear it later
```
Comment (WingLim) [inline site/docs/plugins/guide.md:113]: Add period.
```suggestion
    // Save the handle of the set interval so we can clear it later.
```

**Review by WingLim — CHANGES_REQUESTED** [review]

**Comment by roj1512** [comment]
About the failed format in the latest commit: I don't get that when I format locally. Maybe it is because of the Deno version?

File: site/docs/zh/plugins/guide.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+好吧，你错过了好东西！
+但你也可以只为 Node.js 写你的插件。
+你仍然可以发布插件，让它作为第三方插件列在这个网站上。
+在这种情况下，你可以使用任何你喜欢的文件夹结构（只要它像其他 npm 项目一样组织结构）。
+只需要通过 `npm install grammy` 安装 grammy，然后开始写代码。
+
+## 如何提交？
+
+If you have a plugin ready, you may simply submit a pull request on GitHub (according to the [Rules of Contribution](#rules-of-contribution)), or notify us in the [community chat](https://t.me/grammyjs) for further help.
```
Comment (roj1512) [inline site/docs/zh/plugins/guide.md:210]: Did you miss translating here?

Comment (WingLim) [reply site/docs/zh/plugins/guide.md]: Sorry, forgot to remove it.

**Comment by WingLim** [comment]
> About the failed format in the latest commit: I don't get that change when I format locally. Maybe it is because of the Deno version of the formatting workflow?

The CI is pass in #162 , why failed in here?

**Review by WingLim — APPROVED** [review]

**Comment by roj1512** [comment]
@WingLim no idea. But can you try making the change the CI error requested manually?

**Comment by WingLim** [comment]
@roj1512 The problem is I don't know why it failed.
And the most wired thing is, the error line did not show up in this PR, it's in #162 or in the main branch.

**Comment by roj1512** [comment]
@WingLim See the details here: https://github.com/grammyjs/website/runs/4568515512?check_suite_focus=true

**Comment by roj1512** [comment]
@WingLim Try making this change manually.
image

**Comment by WingLim** [comment]
@roj1512 But this line did not exist in this branch...

**Comment by roj1512** [comment]
Thanks, @WingLim!

**Pull request merged**